### ISSUE TITLE
AppImage

### ISSUE THREAD

**Issue by izzylaif** [paolostivanin/OTPClient#280]
Could you also release official AppImage, since you already do flatpaks?
AppImage seems to be superior in so many ways, including portability.

**Comment by paolostivanin** [comment]
Hi, it depends on how much this is gonna require for creating and maintaining it. Will look into it though.

Thanks :)

**Comment by izzylaif** [comment]
Great news. The creator and main maintainer of appimage format is very keen to onboard new adopters of the packaging, so he probably will help. After the initial setup it is mostly an automated process for future releases.

The pros of appimages is they don't need any infrastructure in the system to work, as opposed to snaps of flatpaks downloading gigabytes worth of packaging ecosystem. Appimages just start on most distros, similar to .dmg on Mac.
Also you can easily force appimages to use their settings in their own config folders, making the whole thing portable and movable. Similar to portableapps in Windows.

**Comment by izzylaif** [comment]
I went ahead and created an appimage. It worked with only one hiccup - the application reads otpclient.ui and shortcuts.ui from hard-coded absolute path /usr/share/otpclient so I had to put those files in that path, and the appimage launched.
if you could change that to relative to the application, creating the appimage will be a breeze

**Comment by paolostivanin** [comment]
Sure, it shouldn't be an issue. A couple of questions:
1. what's the best path for an Appimage for the UI files?
2. can the Appimage read the `~/.config` folder or is it like flatpak?

**Comment by izzylaif** [comment]
I've attached the compiled one.
otpclient.AppImage -> that the executable

 If these folders are present near the executable: 
 
 otpclient.AppImage.home  
  otpclient.AppImage.config 

the appimage will use those instead of /home/user/ and /home/user/.config
making the application portable. You can out it in a thumbdrive and use on any machine with all your settings.
If they folders are not present, the appimage will use the user home folder instead.

If you run the appimage, it will unpack to /tmp/ i.e. /tmp/.mount_otpcliXYZ. The XYZ changes everytime.
In this folder, the *.ui are actually present in /tmp/.mount_otpclioXYZ/usr/share/otpclient and the variables are exported by AppRun file (inside the appimage or the /tmp when run).
However otcpclient seems to want them in absolute path. I believe you have $PREFIX set to root/usr or something like that.
So basically all that is needed to do is change that absolute path to relative to the application.
Currently, to run the appimage provided you actually need root/usr/share/otpclient/*.ui files to exist. If you run the appimage from console without those files in place you will see an error
`Setting $HOME to /mnt/TEST/OTP/otpclient.AppImage.home
Setting $XDG_CONFIG_HOME to /mnt/TEST/OTP/otpclient.AppImage.config

(otpclient:3197): Gtk-ERROR **: 13:53:32.448: failed to add UI: Failed to open file “/usr/share/otpclient/otpclient.ui”: No such file or directory
zsh: trace trap (core dumped)  /mnt/TEST/OTP/otpclient.AppImage
`

Hope that make any sense

[otpclient.zip](https://github.com/paolostivanin/OTPClient/files/10529888/otpclient.zip)

**Comment by paolostivanin** [comment]
I see, thanks! One question: isn't there a recipe (like flatpak and snap do) for building an appimage? 
While I appreciate your effort (really, thanks a lot!), I don't really like to publish binaries built by someone else :) I hope you understand.

**Comment by izzylaif** [comment]
neither do I want to be a maintainer or trust any other packager. that's the whole point why I approached you to generate an official appimage.
The recipes are .yml files. I don't have the yml for otpclient, but here is an example of chrome
[https://github.com/AppImageCommunity/pkg2appimage/blob/master/recipes/Google_Chrome.yml](url)

**Comment by paolostivanin** [comment]
Thanks!
I don't really have a lot of free time these days, so I prefer to focus on code rather than on yet another packaging format.
I'll leave this open for contributors though :smile:

**Comment by izzylaif** [comment]
@probonopd we summon you

**Comment by probonopd** [comment]
As @paolostivanin wrote, he prefers to focus on code rather than on "yet another packaging format" (although I wouldn't characterize AppImage as just that).

**Comment by paolostivanin** [comment]
hello, thanks for your help @Tomurisk . I do not understand why you say "for wider compatibility" ? 
The only really breaking change is the update TO version `4.3.0` if someone wants to update FROM a version earlier than `3.1.4` directly.
But doing 3.2.0 -> 4.0.0 -> 4.4.1 should be completely safe.

There were also tons of bug fixes and improvements also to import/export, QR parsing, etc, so sticking with a 2y6m old release is not really the best approach.

Thank you though 😄

**Comment by Tomurisk** [comment]
@paolostivanin Hi, first of all thanks for your response and for creating OTPClient. When I mean "wider compatibility", I mean older glibc (in this case 2.34), which still allows users to run OTPClient on reasonably older systems, like Ubuntu 22.04, Debian 12, RHEL 9 and the like. Within my experimentation, I technically got my build script for the AppImage to run on Debian 11 (glibc 2.31), but besides Ubuntu 20.04, Debian 11, and Leap 15.5, there's not a wide spectrum for additional distros (+ Debian 11 support ends in like two months or so). I'm pretty sure that it's possible to  build it on something as old as Ubuntu 18.04, however the major roadblock is GTK version. It's indeed perfectly possible to build the latest upstream version on something like Arch, Rawhide or Tumbleweed, but this sort of defeats the whole point of AppImage, because it's no longer portable. There's this pseudo-runtime builder called [sharun](https://github.com/VHSgunzo/sharun), but it takes a chunk of your system by tracing runtime libraries, bundles into AppImage and while in theory it should run on something as old as Ubuntu 14.04, in practice it also bloats up the AppImage, making Flatpak the more reasonable choice for this type of application.

So there's this dilemma: do we want portability, or do we want the latest version? From what I see, you're into newest features and you tend to update dependencies liberally, so I doubt that even if I did manage to craft an AppImage pipeline to match the best of two worlds (Ubuntu 18.04-era glibc + latest upstream release), it'd be hard to lock the AppImage dependencies to certain versions, because with fast update cycle, the AppImage pipeline would break rapidly, it'd be in need of extensive maintenance, as well as that it may take 45 minutes to build GTK with the entire trail of dependencies. I could argue that many AppImage maintainers know that this is quite a big deal, and we have a fair share of AppImages on the market that get deployed but don't actually run once you download, because they naturally get broken by updates... Therefore, I don't believe that a well-maintained AppImage is possible for OTPClient in the long run, unless OTPClient gets split into a conservative LTS codebase (which I don't think you will do, nor do I think you have to 😄)

But yeah, that's the whole problem. The 3.2.0 AppImage I hacked up works just fine, so consider that it was a great thing long before now 🫡

**Comment by paolostivanin** [comment]
Hi! First off, a huge thank you for the effort you've put into creating and maintaining the AppImage so far. That 3.2.0 build was a great contribution, and I really appreciate the "hackery" it took to get it running 😄 
I completely hear you on the maintenance burden. The struggle between "building on ancient glibc" and "using modern  features" is a notorious headache in the Linux ecosystem. Honestly, I’m fine with whatever level of support you’re able to provide! Please don't feel pressured to maintain a complex pipeline if it becomes a second full-time job.
To give you a bit of insight into how I handle dependencies: I actually try to target Debian Stable as my baseline when bumping versions. For example, with the v5 release, all dependencies should be compatible with the current Debian Stable release.
I’m definitely not trying to chase the bleeding edge just for the sake of it, but I do prefer the features provided by newer libraries over supporting systems that are nearing EOL. If that makes the AppImage too difficult to maintain in the long run, I totally understand - Flatpak is indeed a very solid alternative for this exact reason.

Thanks again for the support and for the very honest feedback!

**Comment by pepa65** [comment]
> So there's this dilemma: do we want portability, or do we want the latest version?

Why not both? Build them with the version number of OTPClient in the name. I would be interested in a version with the libraries that are not available with Ubuntu 24.04.

**Comment by probonopd** [comment]
It's all about tradeoffs: You can build on the latest and greatest, but then you have to bundle _everything_ (e.g., with go-appimage or sharun). Or you change your software to need less recent dependencies...
